crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,275 @@
+from enum import Enum, auto
+
+class Colors(Enum):
+    RED = auto()
+    BLUE = auto()
+    YELLOW = auto()
+
+def f(var: X):
+    ...  # the type of `var` here is `X` (which is equivalent to `Literal[Colors.RED, Colors.BLUE, Colors.YELLOW]`)
+
+    if var is Colors.RED:
+        ...  # var is narrowed to `Literal[Colors.RED]`
+    else:
+        ...  # var is narrowed to `Literal[Colors.BLUE, Colors.YELLOW]`
+```
+
+## Single-value types
+
+For a given type `T`, `T` can be said to be a single-value type if there exists an
+[equivalence relation](https://en.wikipedia.org/wiki/Equivalence_relation)
+between all inhabitants of the type and the equivalence relation is satisfied between
+all inhabitants of the type.
+
+For a Python type `T` to be categorised as a single-value type:
+
+- All inhabitants of `T` must be instances of the same runtime class `U`
+- `U` must have `__eq__` and `__ne__` methods such that the equality relation between instances of `U`
+    is reflexive, symmetric, and transitive.
+- For any two inhabitants of `T` named `x` and `y`, `x == y`.
+
+Nearly all singleton types are single-value types, but the reverse is not true.
+Many single-value types exists that are not singleton types: examples include
+`Literal[123456]`, `Literal[b"foo"]`, `tuple[Literal[True], Literal[False]]`, and `Literal["foo"]`.
+All runtime inhabitants of the `Literal["foo"]` type are equal to the string `"foo"`.
+However, they are not necessarily the *same* object in terms of identity;
+multiple `str` instances equal to `"foo"` can exist at runtime at different memory addresses.[^1]
+This means that it is not safe to narrow a single-value type by identity unless it is also known
+that the type is a singleton type.
+
+Single-value types in unions can be safely narrowed using inequality (but not using equality):
+
+```py
+from typing import Literal
+
+def f(x: str | Literal[1]):
+    if x != 1:
+        ...  # `x` can be narrowed to `str` if not equal to `1`
+    else:
+        ...  # type of `x` is still `str | Literal[1]`
+             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
+             # to compare equal with `1`)
+
+    if x == 1:
+        ...  # type of `x` is still `str | Literal[1]`
+             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
+             # to compare equal with `1`)
+    else:
+        ...  # `x` can be narrowed to `str` if not equal to `1`
+```
+
+An example of a singleton type that is not a single-value type
+would be `Literal[Ham.A]` for the following `Ham` enum:
+because the `Ham` class overrides `__eq__`, we can no longer say for sure
+that the `Ham` type can be safely narrowed using inequality as in the above example
+(but we *can* still narrow the `Ham` type using identity, as with all singleton types):
+
+```py
+from enum import Enum, auto
+
+class Ham(Enum):
+    A = auto()
+    B = auto()
+    C = auto()
+
+    def __eq__(self, other):
+        return True
```

Comment:
It may be confusing to readers that technically `Literal[Ham.A]` still is a single-value type, given this definition of `Ham`, because all `Ham` instances compare equal to everything! And this is technically still a valid equivalence relation, in which all inhabitants of `Literal[Ham.A]` do compare equal.

I think in  _practice_ though, we will not want to give red-knot that level of intelligence about the specific implementation of `__eq__`; we will rather want to just say that if `__eq__` is defined, we don't understand the equality relation, and can no longer conclude that `Literal[Ham.A]` is a single-valued type.

It may be useful to clarify this distinction, or it may be fine to just change this `return True` to `return False` and avoid the issue altogether here.